### ISSUE TITLE
Lock wait timeout in MySQL when mixing `int` and `uint` in where clause of `UpdateAsync()`

### ISSUE THREAD

**Issue by dabide** [linq2db/linq2db#3726]
### Describe your issue
A simple update causes a lock wait timeout when mixing `int` and `uint` in the where clause.
```
MySqlConnector.MySqlException (0x80004005): Lock wait timeout exceeded; try restarting transaction
   at MySqlConnector.Core.ServerSession.ReceiveReplyAsyncAwaited(ValueTask`1 task) in /_/src/MySqlConnector/Core/ServerSession.cs:line 960
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 43
   at MySqlConnector.MySqlDataReader.ActivateResultSet(CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 136
   at MySqlConnector.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 465
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 56
   at MySqlConnector.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlCommand.cs:line 296
   at LinqToDB.Data.DataConnection.ExecuteNonQueryAsync(CancellationToken cancellationToken)
   at LinqToDB.Data.DataConnection.ExecuteNonQueryDataAsync(CancellationToken cancellationToken)
   at LinqToDB.Data.DataConnection.ExecuteNonQueryDataAsync(CancellationToken cancellationToken)
   at LinqToDB.Data.DataConnection.QueryRunner.ExecuteNonQueryAsync(CancellationToken cancellationToken)
   at LinqToDB.Linq.QueryRunner.NonQueryQueryAsync(Query query, IDataContext dataContext, Expression expression, Object[] ps, Object[] preambles, CancellationToken cancellationToken)
   at LinqToDB.Linq.QueryRunner.NonQueryQueryAsync(Query query, IDataContext dataContext, Expression expression, Object[] ps, Object[] preambles, CancellationToken cancellationToken)
   at LinqToDB.Linq.ExpressionQuery`1.LinqToDB.Async.IQueryProviderAsync.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at LinqToDB.LinqExtensions.UpdateAsync[T](IUpdatable`1 source, CancellationToken token)
```

### Steps to reproduce

1. Create a table with an unsigned int column `bar`
2. Run the following query where the `bar` variable is an `int`:
```c#
int bar = 1234;
await db.Foo.Where(f => f.Bar == bar)
    .Set(f => f.Baz, "Baz")
    .UpdateAsync();
```
This generates the following SQL:
```sql
SET     @Bar = 1234;
SET     @Baz = 'Baz';

UPDATE
	`foo` `foo_1`
SET
	`foo_1`.`baz` = @Baz
WHERE
	Cast(`foo_1`.`bar` as SIGNED) = @Bar
```

I suspect that the `Cast()` makes MySQL try to lock the entire table, which in our case times out in production.

If LINQ to DB would have generated the following SQL, it would have worked:
```sql
SET     @Bar = 1234;
SET     @Baz = 'Baz';

UPDATE
	`foo` `foo_1`
SET
	`foo_1`.`baz` = @Baz
WHERE
	`foo_1`.`bar` = Cast(@Bar as UNSIGNED)
```

Is there any reason that the columns are casted instead of the parameters?

### Environment details

`Linq To DB` version: 4.1.0

Database (with version): MySQL 5.7

ADO.NET Provider (with version): MySqlConnector 2.1.10

Operating system: Any

.NET Version: 6

**Comment by MaceWindu** [comment]
Not tests, but it looks like C# compiler emits convert on unsigned argument of == operator.

We probably could do something about it

**Comment by MaceWindu** [comment]
As workaround specify type cast explicitly on `bar`

**Comment by dabide** [comment]
@MaceWindu The trouble here is finding existing code that uses these comparisons. I used ReSharper pattern search, and found 64 places with `someUint32Expression == someInt32Expression`. Then you have the other way around, then other operators, then `array.Contains(...)` and so on...

Some places, this would "only" cause slow performance, not a lock timeout. I really hope this can be fixed on the linq2db side.

BTW, is the behavior where the C# compiler emits convert new in .NET 6? The code that caused the timeout has been unchanged for a long time, and used to work. However, we updated from .NET Core 3.1 to .NET 6 recently.
